internal/gitinterface/references.go

```diff
@@ -0,0 +1,164 @@
+
+package gitinterface
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+const (
+	RefPrefix       = "refs/"
+	BranchRefPrefix = "refs/heads/"
+	TagRefPrefix    = "refs/tags/"
+	RemoteRefPrefix = "refs/remotes/"
+)
+
+var (
+	ErrReferenceNotFound = errors.New("requested Git reference not found")
+)
+
+// GetReference returns the tip of the specified Git reference.
+func (r *Repository) GetReference(refName string) (Hash, error) {
+	stdOut, stdErr, err := r.executeGitCommand("rev-parse", refName)
+	if err != nil {
+		if strings.Contains(stdErr, "unknown revision or path not in the working tree") {
+			return ZeroHash, ErrReferenceNotFound
+		}
+		return ZeroHash, fmt.Errorf("unable to read reference '%s': %s", refName, stdErr)
+	}
+
+	hash, err := NewHash(strings.TrimSpace(stdOut))
+	if err != nil {
+		return ZeroHash, fmt.Errorf("invalid Git ID for reference '%s': %w", refName, err)
+	}
+
+	return hash, nil
+}
+
+// SetReference sets the specified reference to the provided Git ID.
+func (r *Repository) SetReference(refName string, gitID Hash) error {
+	_, stdErr, err := r.executeGitCommand("update-ref", "--create-reflog", refName, gitID.String())
+	if err != nil {
+		return fmt.Errorf("unable to set Git reference '%s' to '%s': %s", refName, gitID.String(), stdErr)
+	}
+
+	return nil
+}
+
+// CheckAndSetReference sets the specified reference to the provided Git ID if
+// the reference is currently set to `oldGitID`.
+func (r *Repository) CheckAndSetReference(refName string, newGitID, oldGitID Hash) error {
+	_, stdErr, err := r.executeGitCommand("update-ref", "--create-reflog", refName, newGitID.String(), oldGitID.String())
+	if err != nil {
+		return fmt.Errorf("unable to set Git reference '%s' to '%s': %s", refName, newGitID.String(), stdErr)
+	}
+
+	return nil
+}
+
+// GetSymbolicReferenceTarget returns the name of the Git reference the provided
+// symbolic Git reference is pointing to.
+func (r *Repository) GetSymbolicReferenceTarget(refName string) (string, error) {
+	stdOut, stdErr, err := r.executeGitCommand("symbolic-ref", refName)
+	if err != nil {
+		return "", fmt.Errorf("unable to resolve %s: %s", refName, stdErr)
+	}
+
+	return strings.TrimSpace(stdOut), nil
+}
+
+// AbsoluteReference returns the fully qualified reference path for the provided
+// Git ref.
+func (r *Repository) AbsoluteReference(target string) (string, error) {
+	if strings.HasPrefix(target, RefPrefix) {
+		return target, nil
+	}
+
+	if target == "HEAD" {
+		return r.GetSymbolicReferenceTarget("HEAD")
+	}
+
+	// Check if branch
+	branchName := BranchReferenceName(target)
+	_, err := r.GetReference(branchName)
+	if err == nil {
+		return branchName, nil
+	}
+	if !errors.Is(err, ErrReferenceNotFound) {
+		return "", err
+	}
+
+	// Check if tag
+	tagName := TagReferenceName(target)
+	_, err = r.GetReference(tagName)
+	if err == nil {
+		return tagName, nil
+	}
+	if !errors.Is(err, ErrReferenceNotFound) {
+		return "", err
+	}
+
+	return "", ErrReferenceNotFound
```

Comment:
How would gittuf be able to handle custom refs?

Reply:
We can't really determine the absolute reference for a shorthand of a custom ref. This is only really useful for when users pass in `main` for `refs/heads/main` and `v1.0.0` for `refs/tags/v1.0.0`.

Reply:
https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-emltrefnamegtemegemmasterememheadsmasterememrefsheadsmasterem documents the behavior of ref resolution - it'd be good if we can match this as close as possible. 

Technically general refs do take precedence over branches. (e.g. given `notes/foo`, `refs/notes/foo` should take priority over `refs/heads/notes/foo`)

Reply:
TIL, very interesting. I've updated this so it checks for just `refs/<name>` first.